### PR TITLE
Use Hid.Net, Convert to Task based Async, and add Android/UWP Samples

### PR DESCRIPTION
**Pull request by MelbourneDeveloper**
**Hid.Net**
Hid.Net is a library which supports multiple C# platforms such as Android, and UWP. It allows for dependency injection at the platform level. This pull request replaces the dependency HidLibrary with Hid.Net. This pull request includes a working Android sample, and a UWP sample that will probably work but has a permissions problem. The intention is for Hid.Net to be refactored to support Linux, and MacOS in future. 

Due to the Task based async nature of Hid.Net all methods and unit tests have been refactored to only cater for Task based async methods. All the NanoS unit tests pass with my Ledger nano, but I cannot test with the other units because I don't have those devices.

To run the Android sample:
-    Plug in the Ledger (I'm using Nano with vendor 11415, pid 1). You might need to change the code for a different Ledger device.
-    Enter your pin and open the Bitcoin app
-    Compile and run the Android sample
-    It should show the firmware version on start up

image

**Breaking changes:**
- All calls to the library must now be awaited
- All Task based async methods have been suffixed with Async
- Existing apps will need new code for connecting to the Ledger Hid device. This is an example of the new connection method:

```
        public async static Task<LedgerClientBase> GetLedgerAsync(LedgerType ledgerType)
        {
            var vid = (ushort)11415;
            var devices = WindowsHidDevice.GetConnectedDeviceInformations();
            var newDevice = devices.Where(d => d.VendorId == vid).ToList()[0];
            var windowsHidDevice = new WindowsHidDevice(newDevice);
            windowsHidDevice.DataHasExtraByte = true;
            await windowsHidDevice.InitializeAsync();
            var ledgerTransport = new HIDLedgerTransport(windowsHidDevice);

            switch (ledgerType)
            {
                case LedgerType.Ledger:
                    return new LedgerClient(ledgerTransport);
                case LedgerType.LegacyLedger:
                    return new LegacyLedgerClient(ledgerTransport);
                case LedgerType.U2F:
                    return new U2FClient(ledgerTransport);
                default:
                    throw new NotImplementedException();
            }
        }
```

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Very cool. Can you fix the spaces and put tab? The diff is huge because of this.

File: LedgerWallet/Serializer.cs [hunk 1]
```diff
@@ -15,34 +10,34 @@ public static byte[] Serialize(KeyPath keyPath)
             Guard.AssertKeyPath(keyPath);
             MemoryStream ms = new MemoryStream();
             ms.WriteByte((byte)keyPath.Indexes.Length);
-            for(int i = 0; i < keyPath.Indexes.Length; i++)
+            for (int i = 0; i < keyPath.Indexes.Length; i++)
             {
                 var bytes = ToBytes(keyPath.Indexes[i], false);
-                ms.Write(bytes, 0, bytes.Length);
+                ms.WriteAsync(bytes, 0, bytes.Length);
```
Comment (NicolasDorier) [inline LedgerWallet/Serializer.cs]: remove Async here, this is in memory stream it is immediate

**Comment by NicolasDorier** [comment]
Can you break this in several, independant pull requests so it is easier to review?

**Comment by MelbourneDeveloper** [comment]
No problem. I will do that. 

I probably won't have time for a couple of days though.

**Comment by NicolasDorier** [comment]
@MelbourneDeveloper awesome, I look forward for this. I am excited it will work everywhere.

**Comment by NicolasDorier** [comment]
Hey, where is your branch?

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier  I'm working on the next pull request. It should be ready in less than an hour.

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier do you want the next pull request with or without the Android and UWP samples?

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier  Oh, sorry, the branch for this pull request has been moved to https://github.com/MelbourneDeveloper/ledger-dotnet-api/tree/ScrappedFullVersion because I'm doing everything again so as to break up the parts in to separate pull requests.

**Pull request closed without merging**